src/rtc_pl031.rs

```diff
@@ -0,0 +1,586 @@
+    /// [`Example` section from `RTC`](struct.RTC.html#example).
+    pub fn new() -> RTC {
+        RTC {
+            // Counts up from 1 on reset at 1Hz (emulated).
+            counter: Instant::now(),
+
+            // The load register is initialized to zero.
+            lr: 0,
+
+            // The match register is initialised to zero (not currently used).
+            // TODO: Implement the match register functionality.
+            mr: 0,
+
+            // The interrupt mask is initialised as not set.
+            imsc: 0,
+
+            // The raw interrupt is initialised as not asserted.
+            ris: 0,
+        }
+    }
+
+    fn get_rtc_value(&self) -> u32 {
+        // Add the counter offset to the seconds elapsed since reset.
+        (self.counter.elapsed().as_secs() as u32) + self.lr
+    }
+
+    /// Handles a write request from the driver at `offset` offset from the
+    /// base register address.
+    ///
+    /// # Arguments
+    /// * `offset` - The offset from the base register specifying
+    ///              the register to be written.
+    /// * `data` - The little endian, 4 byte array to write to the register
+    ///
+    /// # Example
+    ///
+    /// You can see an example of how to use this function in the
+    /// [`Example` section from `RTC`](struct.RTC.html#example).
+    pub fn write(&mut self, offset: u32, data: &[u8; 4]) -> Result<(), Error> {
+        let val = u32::from_le_bytes(*data);
+
+        match offset {
+            RTCDR => (), // Ingore writes to the data register.
+            RTCMR => {
+                // Set the match register, though this is not currently used.
+                // TODO: Implement the match register functionality.
+                self.mr = val;
+            }
+            RTCLR => {
+                // Writing to the load register adjusts both the load register
+                // and the counter to ensure that a write to RTCLR followed by
+                // an immediate read of RTCDR will return the loaded value.
+                self.counter = Instant::now();
+                self.lr = val;
+            }
+            RTCCR => {
+                // Writing 1 to the control register resets the RTC value,
+                // which means both the counter and load register are reset.
+                if val == 1 {
+                    self.counter = Instant::now();
+                    self.lr = 0;
+                }
+            }
+            RTCIMSC => {
+                // Set or clear the interrupt mask.
+                self.imsc = val & 1;
+            }
+            RTCRIS => (), // Ignore writes to the interrupt status register.
+            RTCMIS => (), // Ignore writes to the masked interrupt status register.
+            RTCICR => {
+                // Writing 1 clears the interrupt.
+                self.ris &= !val;
+            }
+            _ => {
+                return Err(Error::BadWriteOffset(offset));
```

Comment:
It looks like there is not much the caller of this function do in case of `BadWriteOffset` other than log the error and increment a metric. For consistency with the `read` function (which handles the same error by doing nothing for now), can we get rid of the return function for this one as well?

When we add metrics to the RTC device we can define two metrics for these error cases:
- bad_read_offset
- bad_write_offset

Reply:
I have updated the `write()` method signature to:
```
pub fn write(&mut self, offset: u16, data: &[u8; 4])
```

I've removed the `Error` enum entirely (as a side-effect, this increases our test coverage).

I've updated tests so that they verify that writes to a read-only or non-existent register have no effect.